abci/proposals/proposals.go

```diff
@@ -48,29 +66,62 @@ func NewProposalHandler(
 	prepareProposalHandler sdk.PrepareProposalHandler,
 	processProposalHandler sdk.ProcessProposalHandler,
 	validateVoteExtensionsFn ve.ValidateVoteExtensionsFn,
+	voteExtensionCodec codec.VoteExtensionCodec,
+	extendedCommitInfoCodec codec.ExtendedCommitCodec,
+	currencyPairStrategy currencypair.CurrencyPairStrategy,
+	metrics servicemetrics.Metrics,
+	opts ...Option,
 ) *ProposalHandler {
-	return &ProposalHandler{
+	handler := &ProposalHandler{
 		logger:                   logger,
 		prepareProposalHandler:   prepareProposalHandler,
 		processProposalHandler:   processProposalHandler,
 		validateVoteExtensionsFn: validateVoteExtensionsFn,
+		voteExtensionCodec:       voteExtensionCodec,
+		extendedCommitCodec:      extendedCommitInfoCodec,
+		currencyPairStrategy:     currencyPairStrategy,
+		metrics:                  metrics,
 	}
+
+	// apply options
+	for _, opt := range opts {
+		opt(handler)
+	}
+
+	return handler
 }
 
 // PrepareProposalHandler returns a PrepareProposalHandler that will be called
 // by base app when a new block proposal is requested. The PrepareProposalHandler
 // will first fill the proposal with transactions. Then, if vote extensions are
 // enabled, the handler will inject the extended commit info into the proposal.
 func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
```

Comment:
I think we need to understand what were the issues Berachain and dYdX had that lead to these wrapped Prepare proposals and changes introduced in prepare proposals.